System.Device.Gpio/GpioController.cs

```diff
@@ -0,0 +1,268 @@
+        }
+
+        private void GpioControllerValueChanged(object sender, Windows.Devices.Gpio.GpioPinValueChangedEventArgs e)
+        {
+            var gpioPinNumber = ((Windows.Devices.Gpio.GpioPin)sender).PinNumber;
+            if ((e.Edge == Windows.Devices.Gpio.GpioPinEdge.FallingEdge) && (_gpioEvents[gpioPinNumber] == PinEventTypes.Falling))
+            {
+                _gpioPinChange[gpioPinNumber].Invoke(this, new PinValueChangedEventArgs(PinEventTypes.Falling, gpioPinNumber));
+            }
+
+            if ((e.Edge == Windows.Devices.Gpio.GpioPinEdge.RisingEdge) && (_gpioEvents[gpioPinNumber] == PinEventTypes.Rising))
+            {
+                _gpioPinChange[gpioPinNumber].Invoke(this, new PinValueChangedEventArgs(PinEventTypes.Rising, gpioPinNumber));
+            }
+        }
+
+        /// <summary>
+        /// Sets the mode to a pin.
+        /// </summary>
+        /// <param name="pinNumber">The pin number in the controller's numbering scheme</param>
+        /// <param name="mode">The mode to be set.</param>
+        public void SetPinMode(int pinNumber, PinMode mode)
+        {
+            if (!IsPinOpen(pinNumber))
+            {
+                throw new IOException($"Pin {pinNumber} needs to be open");
+            }
+
+            // Safe cast, same enum on nanoFramework
+            _gpioPins[pinNumber].SetDriveMode((Windows.Devices.Gpio.GpioPinDriveMode)mode);
+        }
+
+        /// <summary>
+        /// Removes a callback that was being invoked for pin at pinNumber.
+        /// </summary>
+        /// <param name="pinNumber">The pin number in the controller's numbering scheme.</param>
+        /// <param name="callback">The callback method that will be invoked.</param>
+        public void UnregisterCallbackForPinValueChangedEvent(int pinNumber, PinChangeEventHandler callback)
+        {
+            _gpioEvents[pinNumber] = PinEventTypes.None;
+            _gpioPins[pinNumber].ValueChanged -= GpioControllerValueChanged;
+        }
+
+        /// <summary>
+        /// Blocks execution until an event of type eventType is received or a period of
+        /// time has expired.
+        /// </summary>
+        /// <param name="pinNumber">The pin number in the controller's numbering scheme.</param>
+        /// <param name="eventTypes">The event types to wait for.</param>
+        /// <param name="timeout">The time to wait for the event.</param>
+        /// <returns>A structure that contains the result of the waiting operation.</returns>
+        public WaitForEventResult WaitForEvent(int pinNumber, PinEventTypes eventTypes, TimeSpan timeout)
+        {
+            _gpioEvents[pinNumber] = eventTypes;
+            _gpioEventsHappening[pinNumber] = PinEventTypes.None;
+            _gpioPins[pinNumber].ValueChanged += GpioControllerWaitForEvents;
+            DateTime dtTimeout = DateTime.UtcNow.Add(timeout);
+            while (DateTime.UtcNow < dtTimeout)
```

Comment:
This `while` loop is going to cannibalize CPU cycles as it's executing waiting for the exit condition to happen.
Has impact on the overall execution and it's not power friendly (prevents the CPU from going to the lowest possible power mode).

nF CLR has means to deal with this by using a MCU timer and putting the thread on wait until the timestamp condition is meet.

As we are not implementing the native side immediately (or we can do so, if we agree on that) suggest that this method (and related) are set to `NotImplementedException`.

Reply:
I was super hesitant to propose it. I've seen how native handle this with the EventSink. I was thinking of something like this. 
Will just remove this from this implementation then. Very few bindings are using the event anyway.

Reply:
I see. But know that this feature has been requested a few time at our end, so it's good to have it.
Agreed that we can leave it out and wait for the next pass.